### PR TITLE
PTL Changes for Shasta

### PR DESCRIPTION
**Pull request by vstumpf**
Still some things need to be changed in PTL for shasta.

#### Describe Bug or Feature
is_localhost fails on shasta
PBS_PUBLIC_HOST_NAME needs to be set
run_copy needs to handle user-hosts
create_temp_file needs to handle user-hosts
Job.create_script needs to handle user-hosts
proc_utils does not support the shasta platform
\@skipOnShasta needs to be implemented
Some usernames are hardcoded in PTL

#### Describe Your Change
is_localhost should check the hostnames reported, and if they're the server hostname, return `True`
Set PBS_PUBLIC_HOST_NAME on the server conf.
Add user-host specific code to run_copy, create_temp_file, and Job.create_script
Allow the shasta platform to use proc_utils
Implement \@skipOnShasta
Change hard-coded usernames to str(TEST_USER)

#### Link to Design Doc
https://pbspro.atlassian.net/wiki/spaces/PD/pages/1327235075/Support+PTL+on+Shasta

#### Attach Test and Valgrind Logs/Output
soon

### CODE REVIEW

File: test/fw/ptl/utils/pbs_testsuite.py [hunk 1]
```diff
@@ -1126,6 +1131,13 @@ def _revert_pbsconf_server(self, vals_to_set):
                 new_pbsconf["PBS_LOG_HIGHRES_TIMESTAMP"] = "1"
                 restart_pbs = True
 
+            # if shasta, set PBS_PUBLIC_HOST_NAME
+            if server.platform == 'shasta':
+                localhost = socket.gethostname()
+                if new_pbsconf["PBS_PUBLIC_HOST_NAME"] != localhost:
+                    new_pbsconf["PBS_PUBLIC_HOST_NAME"] = localhost
+                    restart_pbs = True
```
Comment (bhroam) [inline test/fw/ptl/utils/pbs_testsuite.py]: Isn't PBS_PUBLIC_HOST_NAME there for the commands?  If it is, we won't need to restart PBS.

File: test/fw/ptl/utils/pbs_testsuite.py [hunk 2]
```diff
@@ -1202,7 +1214,12 @@ def revert_pbsconf(self):
             "PBS_LOG_HIGHRES_TIMESTAMP": None
         }
 
-        self._revert_pbsconf_server(vals_to_set)
+        server_vals_to_set = vals_to_set
+
+        if primary_server.platform == 'shasta':
+            server_vals_to_set["PBS_PUBLIC_HOST_NAME"] = None
```
Comment (bhroam) [inline test/fw/ptl/utils/pbs_testsuite.py]: Why are you setting this to None?  Don't you need it set?

File: test/fw/ptl/utils/pbs_dshutils.py [hunk 3]
```diff
@@ -1933,13 +1933,20 @@ def create_temp_file(self, hostname=None, suffix='', prefix='PtlPbs',
             else:
                 os.write(fd, body)
         os.close(fd)
+
+        if self.get_platform() == 'shasta' and not hostname and asuser:
+            asuser = PbsUser.get_user(asuser)
+            if asuser.host:
+                self.logger.info('set hostname to userhost')
```
Comment (bhroam) [inline test/fw/ptl/utils/pbs_dshutils.py]: I don't find this log message useful by itself.  I'd either delete it, or add the host.

File: test/fw/ptl/lib/pbs_testlib.py [hunk 4]
```diff
@@ -13930,16 +13932,23 @@ def create_script(self, body=None, asuser=None, hostname=None):
...
             self.du = DshUtils()
         # First create the temporary file as current user and only change
         # its mode once the current user has written to it
         fn = self.du.create_temp_file(hostname, prefix='PtlPbsJobScript',
                                       asuser=asuser, body=body)
-
         self.du.chmod(hostname, fn, mode=0755)
-        if not self.du.is_localhost(hostname):
+        # on shasta, if we have already set the hostname,
+        # it does not need to be copied again
```
Comment (bhroam) [inline test/fw/ptl/lib/pbs_testlib.py]: Could you add a why in here?  Have you already copied it?  Maybe say that and where in the code that copied it.

File: test/tests/pbs_smoketest.py [hunk 5]
```diff
@@ -327,6 +327,7 @@ def test_limits(self):
         self.server.expect(JOB, a, attrop=PTL_AND)
 
     @skipOnCpuSet
+    @skip('This test is broken for oss')
```
Comment (bhroam) [inline test/tests/pbs_smoketest.py]: Don't skip this test.  If you do, it will likely not be unskipped when it is fixed.

File: test/tests/functional/pbs_equiv_classes.py [hunk 6]
```diff
@@ -1495,29 +1495,30 @@ def test_server_queue_limit(self):
         # Set queue limit
         a = {
             'max_run': '[o:PBS_ALL=100],[g:PBS_GENERIC=20],\
-                       [u:PBS_GENERIC=20],[g:tstgrp01 = 8],[u:pbsuser1=10]'}
+                       [u:PBS_GENERIC=20],[g:tstgrp01 = 8],[u:%s=10]' %
```
Comment (bhroam) [inline test/tests/functional/pbs_equiv_classes.py]: Since you are fixing usernames, you should fix the group names while you are at it.

File: test/fw/ptl/utils/pbs_dshutils.py [hunk 7]
```diff
@@ -1019,8 +1019,12 @@ def run_cmd(self, hosts=None, cmd=None, sudo=False, stdin=None,
...
             if as_script:
                 # must pass as_script=False otherwise it will loop infinite
-                self.rm(hostname, path=_script, as_script=False,
-                        level=level, runas=runas)
+                if platform == 'shasta' and not runas:
+                    self.rm(hostname, path=_script, as_script=False,
+                            level=level, runas=runas)
+                else:
+                    self.rm(hostname, path=_script, as_script=False,
+                            level=level)
```
Comment (bhroam) [inline test/fw/ptl/utils/pbs_dshutils.py]: OK, this looks wrong.  First you took runas=runas out of the normal case.  Second, you said 'not runas' and then added runas=runas.  Won't it be empty?

File: test/fw/ptl/utils/pbs_dshutils.py [hunk 8]
```diff
@@ -1919,13 +1944,20 @@ def create_temp_file(self, hostname=None, suffix='', prefix='PtlPbs',
             else:
                 os.write(fd, body)
         os.close(fd)
+
+        if self.get_platform() == 'shasta' and not hostname and asuser:
+            asuser = PbsUser.get_user(asuser)
+            if asuser.host:
```
Comment (agrawalravi90) [inline test/fw/ptl/utils/pbs_dshutils.py]: same comment, i don't think this needs to be shasta specific.

File: test/fw/ptl/utils/pbs_dshutils.py [hunk 9]
```diff
@@ -1919,13 +1944,20 @@ def create_temp_file(self, hostname=None, suffix='', prefix='PtlPbs',
...
+                hostname = asuser.host
+
         # if temp file to be created on remote host
         if not self.is_localhost(hostname):
             if asuser is not None:
                 # by default mkstemp creates file with 0600 permission
                 # to create file as different user first change the file
                 # permission to 0644 so that other user has read permission
-                self.chmod(hostname, tmpfile, mode=0644)
+                self.chmod(tmpfile, mode=0644)
```
Comment (agrawalravi90) [inline test/fw/ptl/utils/pbs_dshutils.py]: why did you remove the hostname arg?

File: test/tests/pbs_smoketest.py (hunk 5 above)
Comment (agrawalravi90) [reply test/tests/pbs_smoketest.py]: this is a smoketest, how is it broken?

File: test/fw/ptl/utils/pbs_dshutils.py (hunk 7 above)
Comment (vstumpf) [reply test/fw/ptl/utils/pbs_dshutils.py]: Interesting, this is the commit i force-pushed but the changes still didn't get added. I must not have committed them.
You're right, it should be `if platform == 'shasta' and runas:`

File: test/fw/ptl/utils/pbs_dshutils.py [hunk 10]
```diff
@@ -1146,11 +1156,17 @@ def run_copy(self, hosts=None, src=None, dest=None, sudo=False, uid=None,
                 cmd += copy_cmd
                 if recursive:
                     cmd += ['-r']
+                if (self.get_platform() == 'shasta' and _runas_user and
+                   _runas_user.port):
+                    cmd += ['-P', _runas_user.port]
```
[the remark this answers is not in the document]
Comment (vstumpf) [reply test/fw/ptl/utils/pbs_dshutils.py]: run_cmd()'s port argument is the port option for ssh. This port is for scp.

File: test/fw/ptl/utils/pbs_dshutils.py (hunk 9 above)
Comment (vstumpf) [reply test/fw/ptl/utils/pbs_dshutils.py]: It was a bug, PTL creates a file on the local host, then chmods the file on the remote host? The file hasn't been copied yet.

File: test/fw/ptl/utils/pbs_testsuite.py (hunk 2 above)
Comment (vstumpf) [reply test/fw/ptl/utils/pbs_testsuite.py]: vals_to_set is weird, we use the keys of the dictionary to determine if the conf var needs to be set. It could be replaced with a list, but that's out of scope.

File: test/tests/pbs_smoketest.py (hunk 5 above)
Comment (vstumpf) [reply test/tests/pbs_smoketest.py]: Sorry, this wasn't supposed to be in the PR, just for testing.
It's broken due to #1267

File: test/tests/functional/pbs_equiv_classes.py (hunk 6 above)
Comment (lisa-altair) [reply test/tests/functional/pbs_equiv_classes.py]: I think it would be reasonable to file a bug to fix group names.

Comment (bhroam) [reply test/tests/functional/pbs_equiv_classes.py]: Since it won't take long to fix the group names, I'd rather do it now.

File: test/fw/ptl/lib/pbs_testlib.py (hunk 4 above)
Comment (vstumpf) [reply test/fw/ptl/lib/pbs_testlib.py]: I'm not even sure why this is in the code, create_temp_file should create the job script on the host specified by hostname. We then chmod the file on that host. Then we do another copy? It seems wrong, I'll remove it.

File: test/fw/ptl/utils/pbs_dshutils.py (hunk 3 above)
Comment (vstumpf) [reply test/fw/ptl/utils/pbs_dshutils.py]: Sorry, it was a debug message.

File: test/fw/ptl/utils/pbs_testsuite.py (hunk 1 above)
Comment (vstumpf) [reply test/fw/ptl/utils/pbs_testsuite.py]: In order to set the pbsconfig, restart_pbs needs to be true.

**Comment by vstumpf** [comment]
Please don't merge until I test tomorrow. It is still open for review, however.

File: test/fw/ptl/utils/pbs_dshutils.py (hunk 9 above)
Comment (agrawalravi90) [reply test/fw/ptl/utils/pbs_dshutils.py]: ok, but now you are passing the file path 'tmpfile' into the 'hostname' argument of the function, you should change this to path=tmpfile

File: test/fw/ptl/lib/pbs_testlib.py [hunk 11]
```diff
@@ -13932,11 +13931,12 @@ def create_script(self, body=None, asuser=None, hostname=None):
                     body[i] = " ".join(line_arr)
             body = '\n'.join(body)
 
+        # If the user has a userhost, the job will run from there
+        # so the script should be made there
```
Comment (agrawalravi90) [inline test/fw/ptl/lib/pbs_testlib.py]: This sounds like a pretty generic logic, so can we remove "if self.platform == 'shasta'" from the check below?

File: test/fw/ptl/utils/pbs_dshutils.py [hunk 12]
```diff
@@ -1019,7 +1019,7 @@ def run_cmd(self, hosts=None, cmd=None, sudo=False, stdin=None,
 
             if as_script:
                 # must pass as_script=False otherwise it will loop infinite
-                if platform == 'shasta' and not runas:
+                if platform == 'shasta' and runas:
```
Comment (agrawalravi90) [inline test/fw/ptl/utils/pbs_dshutils.py]: I think we can just pass 'runas' to rm() in either case, did you see any issue doing that in the generic case?

File: test/fw/ptl/utils/pbs_dshutils.py [hunk 13]
```diff
@@ -1156,9 +1154,8 @@ def run_copy(self, hosts=None, src=None, dest=None, sudo=False, uid=None,
                 cmd += copy_cmd
                 if recursive:
                     cmd += ['-r']
-                if (self.get_platform() == 'shasta' and _runas_user and
-                   _runas_user.port):
-                    cmd += ['-P', _runas_user.port]
+                if self.get_platform() == 'shasta' and runas and runas.port:
```
Comment (agrawalravi90) [inline test/fw/ptl/utils/pbs_dshutils.py]: Again, I feel that the shasta check is not needed here, if runas and runas.port is set, I think we should just use it.

File: test/fw/ptl/utils/pbs_dshutils.py [hunk 14]
```diff
@@ -1300,7 +1297,8 @@ def is_localhost(self, host=None):
             self._h2l[host] = True
             return True
         # on a shasta machine, the name returned by `hostname` (pbs-host) is
-        # different than the one we tell PTL to use (pbs-service-nmn).
+        # different than the one we tell PTL to use (pbs-service-nmn). This
```
Comment (agrawalravi90) [inline test/fw/ptl/utils/pbs_dshutils.py]: Just for my understanding, can you please explain why we ask PTL to use a different "local" hostname than the one that `hostname` returns ? Or if this is documented, point me to it? Thanks.

File: test/fw/ptl/utils/pbs_dshutils.py (hunk 9 above)
Comment (agrawalravi90) [reply test/fw/ptl/utils/pbs_dshutils.py]: > ok, but now you are passing the file path 'tmpfile' into the 'hostname' argument of the function, you should change this to path=tmpfile

I think you missed this comment

Comment (vstumpf) [reply test/fw/ptl/utils/pbs_dshutils.py]: Ah, I missed that. Thanks!
